hivemind/utils/mpfuture.py

```diff
@@ -111,120 +102,197 @@ def _set_event_threadsafe(self):
+        if not self.synchronize or os.getpid() == self._origin_pid or self._state in TERMINAL_STATES:
+            return
+
+        self._initialize_backend_if_necessary()
+
+        status_updated = threading.Event()
+        _, existing_status_event = self._status_requests.setdefault(self._uid, (self, status_updated))
+        # this line checks if another thread is synchronizing concurrently, assuming that setdefault to be atomic
+
+        if existing_status_event != status_updated:
+            existing_status_event.wait(MPFuture.HARD_UPDATE_TIMEOUT)
+            return
+
+        # otherwise create a new request for synchronization
+
+        try:
+            with MPFuture._update_lock if self._use_lock else nullcontext():
+                payload = (self._use_lock, self._process_wide_pipe)
+                self._sender_pipe.send((self._uid, MessageType.STATE_REQUEST, payload))
+            status_updated.wait(MPFuture.SOFT_UPDATE_TIMEOUT)
+            if not status_updated.is_set():
+                logger.warning(f"Status update took over {MPFuture.SOFT_UPDATE_TIMEOUT}, expect performance issues")
+                status_updated.wait(MPFuture.HARD_UPDATE_TIMEOUT - MPFuture.SOFT_UPDATE_TIMEOUT)
+                if not status_updated.is_set() and not self.cancel():
+                    with suppress(InvalidStateError, RuntimeError):
+                        self.set_exception(
+                            TimeoutError(
+                                f"Status update took over {MPFuture.HARD_UPDATE_TIMEOUT} seconds, "
+                                f"MPFuture is cancelled"
+                            )
+                        )
+                    status_updated.set()  # this triggers any concurrent _synchronize_if_necessary calls to finish
+        except (ConnectionError, BrokenPipeError, EOFError) as e:
+            logger.error(f"MPFuture was cancelled because sender pipe is broken. Origin process is probably down.")
+            if not self.cancel():
+                with suppress(InvalidStateError, RuntimeError):
+                    self.set_exception(e)
+        finally:
+            self._status_requests.pop(self._uid, None)
 
     def set_result(self, result: ResultType):
-        if os.getpid() == self._origin_pid:
-            super().set_result(result)
-            MPFuture._active_futures.pop(self._uid, None)
-        elif self._state in TERMINAL_STATES:
+        if self._state in TERMINAL_STATES:
             raise InvalidStateError(f"Can't set_result to a future that is {self._state} ({self._uid})")
+        elif os.getpid() == self._origin_pid:
+            MPFuture._active_futures.pop(self._uid, None)
+            self._set_event_if_necessary()
         else:
-            self._state_cache[self._state], self._result = base.FINISHED, result
-            self._send_update(UpdateType.RESULT, result)
+            self._send_update(MessageType.RESULT, result)
+        super().set_result(result)
```

Comment:
nit: In Python 3.7, this may raise `RuntimeError` (instead of `InvalidStateError`) if `set_result` is called concurrently inside one process.

This is a minor issue and we've agreed it won't be fixed.